src/SDK/SmartDetectors.Extensions/Extensions/StringExtensions.cs

```diff
@@ -57,13 +60,25 @@ public static string ToSha256Hash(this string s)
         /// <returns>The evaluated result string</returns>
         internal static string EvaluateInterpolatedString(this string interpolatedStringDefinition, object source)
         {
+            // If this is not an interpolated string - just return it as-is
             if (!interpolatedStringDefinition.Contains("{"))
             {
                 return interpolatedStringDefinition;
             }
 
+            // Code that interprets the string as an interpolated string
             string code = "$\"" + interpolatedStringDefinition + "\"";
-            return CSharpScript.EvaluateAsync<string>(code, globals: source, globalsType: source.GetType()).Result;
+
+            // Get the script from cache (improves performance since the script is already compiled)
+            string key = source.GetType().FullName + "#" + code;
+            if (!InterpolatedStringScripts.TryGetValue(key, out Script<string> script))
+            {
+                script = InterpolatedStringScripts[key] = CSharpScript.Create<string>(code, globalsType: source.GetType());
+            }
```

Comment:
change to concurrent dictionary, GetOrAdd

Reply:
Good point

---
In reply to: [231036400](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/18#discussion_r231036400) [](ancestors = 231036400)